File: leetcode/cpp/sliding_window/lc_3_longest_substring_without_repeating_characters.cpp
#include <string>
#include <unordered_set>

/**
 * https://leetcode.com/problems/longest-substring-without-repeating-characters/description/
 *
 * Time complexity: O(n) where n = length of s
 * Space complexity: O(n) where n = length of s
 */
class Solution {
public:
    int lengthOfLongestSubstring(std::string s) {
        // first handle the edge case if we are given
        // a string with less than 2 chars
        // simply return it's size
        if (s.size() < 2) {
            return s.size();
        }

        // create our seen set to keep track of the unique
        // chars in the substring (our sliding window)
        std::unordered_set<char> seen;

        // init the max length we found so ar
        int maxLength = 0;
        // init the start index i of our sliding window
        int i = 0;

        // loop through all the chars in the string
        for (const char& chr: s) {
            // as long as we have seen this char before
            // keep removing it from our seen set
            // and increment the start of our sliding window
            // until we no longer have this char in our seen set
            while (seen.contains(chr)){
                seen.erase(s[i]);
                ++i;
            }
            // finally, insert the current char into our seen set
            seen.insert(chr);

            // keep track of the max length of our seen set
            // which is our substring
            maxLength = std::max(static_cast<int>(seen.size()), maxLength);
        }

        // finally, return the max length we have stored so far.
        return maxLength;
    }
};

File: leetcode/sliding_window/lc_3_longest_substring_without_repeating_characters.py
'''
https://leetcode.com/problems/longest-substring-without-repeating-characters/description/
'''
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        # handle edge case where 
        # we have an empty string
        if not s:
            return 0

        # init our max_length variable
        # to 1 because at the least
        # we can return a substring of 1 char
        max_length = 1

        # init our first pointer that
        # points to the start of a unique
        # substring
        start = 0

        # we will use a set to keep track 
        # of chars we have seen
        # initially add the first char
        # to it
        seen = set([s[0]])

        # we will loop through all indexes
        # after the first index 0
        for end in range(1, len(s)):
            # check if we have seen this char
            # before (i.e., is it in our seen set)
            # if it is, then loop until it is not
            while s[end] in seen:
                # on each iteration, remove the
                # current char pointed too by start
                # pointer from the seen set
                seen.remove(s[start])
                # and move start one index forward
                # this will remove chars until we
                # remove the duplicate, thus keeping
                # a unique substring
                start += 1
            
            # add the char to our set
            # this will do nothing if it already exists
            seen.add(s[end])

            # calculate the substring length
            length = end-start+1
            # and update our max length with it
            max_length = max(length, max_length)

        # finally we will return the max length
        return max_length
